### PR TITLE
Platform Displacement 2.1

### PR DESCRIPTION
**Pull request by arthurtilly**
A few fixes to PD2.

* Makes PD2 default and removes the config define due to the increased stability.
* Fixes bug where jumping off a platform immediately after only being on it for one frame would not apply inertia.
* Fixes bug where ledge-grabbing a laterally moving platform would result in Mario's position being offset.
* Changes OBJ_FLAG_VELOCITY_PLATFORM to OBJ_FLAG_NO_AUTO_DISPLACEMENT
* * This flag now completely overrides the transform displacement rather than just adding to it. This means it can now be used to disable platform displacement entirely on an object as well as being used for TTC treadmills.
* * Gave the flag to the JRB sliding ship box as in vanilla it doesn't displace Mario either.
* Simplified a lot of the PD2 code to take better advantage of the new mathutils

### CODE REVIEW

File: src/game/platform_displacement.c [hunk 1]
```diff
@@ -56,50 +64,32 @@ void update_mario_platform(void) {
...
+	vec3_copy(displaceInfo->prevPos, pos);
+
+	// Update yaw
+	vec3_set(yawVec, sins(yaw), 0, coss(yaw));
+	linear_mtxf_transpose_mul_vec3(*platform->header.gfx.throwMatrix, displaceInfo->prevTransformedYawVec, yawVec);
+	displaceInfo->prevYaw = yaw;
+
+	// Update platform and timer
+	displaceInfo->prevPlatform = platform;
+	displaceInfo->prevTimer = gGlobalTimer;
```
Comment (gheskett) [inline src/game/platform_displacement.c:85]: Tabs

File: src/game/platform_displacement.c [hunk 2]
```diff
@@ -109,64 +99,39 @@ void apply_platform_displacement(struct PlatformDisplacementInfo *displaceInfo,
...
+        }
     }
 
-    // Transform from world positions to relative positions for use next frame
-    linear_mtxf_transpose_mul_vec3f(*platform->header.gfx.throwMatrix, scaledPos, pos);
-    scale_vec3f(displaceInfo->prevTransformedPos, scaledPos, platform->header.gfx.scale, TRUE);
-    vec3f_add(pos, platformPos);
+    Vec3f oldPos;
+	vec3_sum(oldPos, displaceInfo->prevPos, posDifference);
+	update_platform_displacement_info(displaceInfo, pos, *yaw, platform);
```
Comment (gheskett) [inline src/game/platform_displacement.c:129]: Tabs

File: src/game/platform_displacement.h [hunk 3]
```diff
@@ -6,24 +6,19 @@
...
-	};
-#endif
+
+struct PlatformDisplacementInfo {
+	Vec3f prevPos;
+	Vec3f prevTransformedPos;
+	Vec3f prevTransformedYawVec;
+	s16 prevYaw;
+	struct Object *prevPlatform;
+	u32 prevTimer;
```
Comment (gheskett) [inline src/game/platform_displacement.h:16]: Tabs

**Review by gheskett — APPROVED** [review]

**Review by gheskett — APPROVED** [review]

**Review by gheskett — CHANGES_REQUESTED** [review]
fix it smh

File: src/game/platform_displacement.c [hunk 4]
```diff
@@ -69,6 +69,9 @@ void update_mario_platform(void) {
 void update_platform_displacement_info(struct PlatformDisplacementInfo *displaceInfo, Vec3f pos, s16 yaw, struct Object *platform) {
     Vec3f scaledPos, yawVec, localPos;
 
+    // Avoid a crash if the platform unloaded its collision while stood on or is static
+    if (platform->header.gfx.throwMatrix == NULL) return;
```
Comment (gheskett) [inline src/game/platform_displacement.c:73]: Ehh...I'm not a fan of inlining returns like this unless they're part of condensed switch blocks. It makes them easy to miss even if they're here as a safeguard.

**Review by gheskett — APPROVED** [review]

File: src/game/platform_displacement.c (hunk 4 above)
Comment (arthurtilly) [reply src/game/platform_displacement.c:73]: this is how it already was in the other function the entire time...

Comment (gheskett) [reply src/game/platform_displacement.c:73]: This may be true and I probably just missed it before but I still don't like it lol

**Pull request merged**